### PR TITLE
Development

### PR DESCRIPTION
**Pull request by inno-asiimwe**
Complete challenge 3

### CODE REVIEW

File: app/auth/views.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        response = {
+            'message':'Successfully registered!',
+            'status':'Success'
+        }
+        return make_response(jsonify(response)), 201
+    response = {
+        'message': 'Failed to register, duplicate user',
+        'status': 'Failed!!'
+    }
+    return make_response(jsonify(response)), 202
```
Comment (patlub) [inline app/auth/views.py]: Status code for duplicate user should be 409(Conflict)

File: app/auth/views.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        response = {
+            'message': 'Successfully changed password',
+            'status': 'Success'
+        }
+        return make_response(jsonify(response)), 200
+    response = {
+        'message': 'Failed to reset password, bad username or password',
+        'status': 'Failed'
+    }
+    return make_response(jsonify(response)), 400
```
Comment (patlub) [inline app/auth/views.py]: should be 401. 400 is for bad request
https://en.wikipedia.org/wiki/List_of_HTTP_status_codes

File: app/bucketlist/views.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+     """
+    my_bucketlist = Bucketlist.query.filter_by(id=b_id).first()
+
+    if not my_bucketlist:
+        abort(404)
+    elif request.method == 'DELETE':
+        my_bucketlist.delete()
+        response = {
+            'status': 'Success'
+        }
```
Comment (patlub) [inline app/bucketlist/views.py]: When you delete a bucket, are the items in that bucket also deleted???

File: app/bucketlist/views.py [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+        return make_response(jsonify(response)), 200
+    elif request.method == 'PUT':
+        my_bucketlist.name = request.data['name']
+        my_bucketlist.description = request.data['description']
+        my_bucketlist.save()
```
Comment (patlub) [inline app/bucketlist/views.py]: Did you account for trying to update a bucket name to one that already exists?

File: app/bucketlist/views.py [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+                    'name': new_item.name,
+                    'description': new_item.description,
+                    'bucketlist_id': new_item.bucketlist_id
+                }
+                return make_response(jsonify(response)), 201
+            response = {
+                'status': 'Failed',
+                'message': 'Item already exists'
+            }
+            return make_response(jsonify(response)), 202
```
Comment (patlub) [inline app/bucketlist/views.py]: 409

File: app/bucketlist/views.py [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+            description: "success"
+        404:
+            description: "Failed"
+    """
+    my_item = Item.query.filter_by(bucketlist_id=b_id, id=i_id).first()
+    if my_item:
+        if request.method == 'PUT':
+            my_item.name = request.data['name']
+            my_item.description = request.data['description']
+            my_item.save()
```
Comment (patlub) [inline app/bucketlist/views.py]: account for updating item name to one that already exists.

File: app/models.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+
+    def password_is_valid(self, password):
+        """Method validates password against its harsh"""
+        return Bcrypt().check_password_hash(self.password, password)
+
+    def encode_auth_token(self, user_id):
+        """Generates an authentication token"""
+        try:
+            payload = {
+                'exp':datetime.datetime.utcnow() + datetime.timedelta(seconds=5),
```
Comment (patlub) [inline app/models.py]: Does the token expire after only 5 seconds

File: app/models.py [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+        """A representation for an instance of a bucketlist"""
+        return '<Bucketlist: %r>' %(self.name)
+
+class Item(db.Model):
+    """Class to define the Items table"""
+    __tablename__ = 'items'
+    id = db.Column(db.Integer, primary_key=True)
+    name = db.Column(db.String(256), nullable=False, unique=True)
+    description = db.Column(db.Text)
+    bucketlist_id = db.Column(db.Integer, db.ForeignKey(Bucketlist.id))
```
Comment (patlub) [inline app/models.py]: You might need a cascade delete property to automatically delete item when bucket is deleted

File: cover/app___init___py.html [hunk 9]
```diff
@@ -0,0 +1,135 @@
+
+
+
+<!DOCTYPE html>
+<html>
+<head>
+    <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
+    
+
```
Comment (patlub) [inline cover/app___init___py.html]: remove unnecessary blank lines

File: app/auth/views.py [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+
+@auth_blueprint.route('/logout', methods=['POST'])
+@auth_required
+def logout_user(resp, auth_token):
+    """Log out a user
```
Comment (patlub) [inline app/auth/views.py]: docstring could have params too
`:param resp:`
`:param auth_token:`
`:return:`

**Comment by inno-asiimwe** [comment]
thanks @patlub for the feedback let me work on all and update repo

**Pull request merged**